### DISCUSSION TITLE
How to completely remove the Terminal?

### DISCUSSION THREAD

**Discussion by rubenanapu** [eclipse-theia/theia#9190] [Q&A] [answered]
### Description

I use Theia on Docker and I want to hide that Terminal menu. I'm only interested in the Code Editor, not on the shell.

In theory, I could only remove "@theia/terminal" from [examples/browser/package.json](https://github.com/theia-ide/theia/blob/master/examples/browser/package.json#L54
), but that doesn't work.

**OS and Theia version:**
Theia 0.8.0
OS: Linux Ubuntu 18.04.2 LTS

**Comment by akosyakov** [comment]
> In theory, I could only remove "@theia/terminal" from examples/browser/package.json, but that doesn't work.

Can you elaborate? Maybe another extension pulls the terminal extension transitively? Please run `yarn why @theia/terminal` and remove all extensions which pull it.

**Comment by kittaakos** [comment]
`@theia/debug` pulls it:

https://github.com/theia-ide/theia/blob/7fe439a30d53846a82d6b0c916de5a69b9da2c77/packages/debug/package.json#L16

**Comment by rubenanapu** [comment]
I'm not using `@theia/debug` @kittaakos, but thanks for the answer.

@akosyakov, with `yarn why @theia/terminal` I found which package uses it. It's used by `@theia/task` that is inherited by `@theia/cpp`:

>  - "_project_#@theia#my-example#@theia#cpp#@theia#task" depends on it

The thing is that I don't want to drop `@theia/cpp` support.

I'll dive deeper into `@theia/terminal` to see how to disable the browser directly there.

Thanks guys.

I'll let you know in case I manage to remove it from the browser.

**Comment by rubenanapu** [comment]
Guys, if I can't easily remove the `@theia/terminal` because I need `@theia/cpp`, does anybody know how to just hide the `Terminal` menu from the web browser?

**Comment by akosyakov** [comment]
You could try to create a custom extension which rebinds terminal contribution and makes it no-op.

**Reply by ballcoach12** [reply]
I'll chime in and add that I cannot seem to make this work either. No matter which approach I try, the Terminal and Terminal->New menus persist. I've tried completely removing the @theia/terminals dependency from the browser and electron app package.json and writing a contribution that returns a no-op on registerMenus() and registerCommands(). 

I'm not too sure how to proceed at this point in order to get this accomplished. Has anyone gotten this to work? If so, would you be so kind as to post a complete code example?

**Reply by rubenanapu** [reply]
What I did was basically hide the **Terminal** menu using JavaScript.

In the end, the index.html would be something like:

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <script type="text/javascript" src="./bundle.js" charset="utf-8"></script>
</head>
<body>
  <div class="theia-preload"></div>

    <script type="text/javascript" >
    document.addEventListener("DOMContentLoaded", function (event) {
        var hideTerminalMenu = setInterval(function () {
            var xMenuItems = document.getElementsByClassName("p-MenuBar-item");
            var xTermMenu = xMenuItems[6];
            if(xTermMenu !== undefined){
                xTermMenu.style.width = '1px';
                xTermMenu.style.padding = '0px';
                clearInterval(hideTerminalMenu);
            }
        }, 1000);
    });
    </script>
</body>
</html>

```

**Comment by datou0412** [comment]
`const menu = this.menuModelRegistry.getMenu(MAIN_MENU_BAR);`
 ` menu.removeNode('7_terminal'); // remove the whole submenu,  wont work`

 ` const menu = this.menuModelRegistry.getMenu(TerminalMenus.TERMINAL_NEW);`
`menu.removeNode(TerminalCommands.NEW.id); // remove one submenu child, it works`

Is this a bug? @akosyakov

**Comment by datou0412** [comment]
> You could try to create a custom extension which rebinds terminal contribution and makes it no-op.

rebind `TerminalFrontendContribution`, and rebind `TaskFrontendContribution` as well

extends the `registerMenus` method to noop
`registerMenus(): void {}`

as well as `registerCommands` , if u want to disabled the command

**Comment by akosyakov** [comment]
@datou0412 A new menu generated each time when you call `getMenu`. You need to call `unregister` to remove terminal menus from the registry at all

**Comment by JonasHelming** [comment]
This might be a good solution for this issue, too:
https://github.com/eclipse-theia/theia/issues/9069

**Comment by rubenanapu** [comment, marked as answer]
What I did was basically hide the Terminal menu using JavaScript.

In the end, the index.html would be something like:

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <script type="text/javascript" src="./bundle.js" charset="utf-8"></script>
</head>
<body>
  <div class="theia-preload"></div>

    <script type="text/javascript" >
    document.addEventListener("DOMContentLoaded", function (event) {
        var hideTerminalMenu = setInterval(function () {
            var xMenuItems = document.getElementsByClassName("p-MenuBar-item");
            var xTermMenu = xMenuItems[6];
            if(xTermMenu !== undefined){
                xTermMenu.style.width = '1px';
                xTermMenu.style.padding = '0px';
                clearInterval(hideTerminalMenu);
            }
        }, 1000);
    });
    </script>
</body>
</html>
```

**Reply by liweijian** [reply]
Hi @rubenanapu  , thanks for you solution. Just wondering if we only need to edit the `src-gen/frontend/index.html` file?

Since I tried to edited it as:

```html
<body>
    <div class="theia-preload"></div>
    <script type="text/javascript" src="./bundle.js" charset="utf-8"></script>

    <script type="text/javascript" >
    alert(1);
    document.addEventListener("DOMContentLoaded", function (event) {
        var hideTerminalMenu = setInterval(function () {
           document.getElementsByClassName('p-MenuBar-item')[6].style.visibility = 'hidden'; 
        }, 1000);
    });
```

And start as: `yarn  start ./workspace --hostname 0.0.0.0 --port 8080`

However, ven the `alert(1)` not work, not sure if anything I'd missed?

**Comment by rubenanapu** [comment]
Hi @liweijian ,

in my case, what I modified was:
* ./dev-packages/application-manager/src/generator/frontend-generator.ts
